parlai/agents/transformer/modules/generator.py

```diff
@@ -27,46 +29,76 @@
     TransformerDecoder,
     TransformerEncoder,
 )
+from parlai.agents.transformer.modules.interfaces import ComponentSpec, TComponent
 from parlai.core.opt import Opt
 from parlai.core.torch_agent import DictionaryAgent
 from parlai.core.torch_generator_agent import TorchGeneratorModel
 from parlai.utils.torch import neginf
 
 
-class TransformerGeneratorModel(TorchGeneratorModel):
+ENCODER_DEFAULT_SPEC = ComponentSpec(TransformerEncoder, TransformerEncoder.Manifest())
+DECODER_DEFAULT_SPEC = ComponentSpec(TransformerDecoder, TransformerDecoder.Manifest())
+
+
+class TransformerGeneratorModel(TorchGeneratorModel, TComponent):
     """
     Implements a full generator model, with one encoder and one decoder.
     """
 
+    @dataclass
+    class Manifest(TComponent.Manifest):
+        encoder: ComponentSpec[TransformerEncoder] = ENCODER_DEFAULT_SPEC
+        decoder: ComponentSpec[TransformerDecoder] = DECODER_DEFAULT_SPEC
+
     @classmethod
     def build_encoder(
-        cls, opt, dictionary, embedding=None, padding_idx=None, reduction_type='mean'
+        cls,
+        opt,
+        dictionary,
+        embedding=None,
+        padding_idx=None,
+        reduction_type='mean',
+        spec: ComponentSpec = ENCODER_DEFAULT_SPEC,
     ):
-        return TransformerEncoder(
+        return spec.klass(
```

Comment:
maybe `spec.encoder.klass(` ?

Reply:
See the `__init__` function below. When `build_encoder` is called, it's passed `manifest.encoder`. And `manifest.encoder` only has `klass` and `manifest`, not `encoder`.